src/Time/StepChoosers/ErrorControl.hpp

```diff
@@ -0,0 +1,345 @@
+ * where \f$h_{\text{new}}\f$ is the new suggested step size \f$h\f$ is the
+ * previous step size, \f$F_{\text{max}}\f$ is the maximum factor by which we
+ * allow the step to increase, \f$F_{\text{min}}\f$ is the minimum factor by
+ * which we allow the step to decrease. \f$F_{\text{safety}}\f$ is the safety
+ * factor on the computed error -- this forces the step size slightly lower
+ * than we would naively compute so that the result of the step will likely be
+ * within the target error. \f$q\f$ is the order of the stepper error
+ * calculation. Intuitively, we should change the step less drastically for a
+ * higher order stepper.
+ *
+ * After the first error calculation, the error \f$E\f$ is recorded, and
+ * subsequent error calculations use a simple PI scheme suggested in \cite
+ * NumericalRecipes section 17.2.1:
+ *
+ * \f[
+ * h_{\text{new}} = h \cdot \min\left(F_{\text{max}},
+ * \max\left(F_{\text{min}},
+ * \frac{F_{\text{safety}}}{E^{0.7 / (q + 1)}
+ *  E_{\text{prev}}^{0.4 / (q + 1)}}\right)\right),
+ * \f]
+ *
+ * where \f$E_{\text{prev}}\f$ is the error computed in the previous step.
+ *
+ * \note The template parameter `ErrorControlSelector` is used to disambiguate
+ * in the input-file options between `ErrorControl` step choosers that are
+ * based on different variables. This is needed if multiple systems are evolved
+ * in the same executable. The name used for the input file includes
+ * `ErrorControlSelector::name()` if it is provided.
+ */
+template <
+    typename EvolvedVariableTag, typename ErrorControlSelector = NoSuchType,
+    typename StepChooserRegistrars = tmpl::list<
+        Registrars::ErrorControl<EvolvedVariableTag, ErrorControlSelector>>>
+class ErrorControl : public StepChooser<StepChooserRegistrars> {
+ public:
+  using evolved_variable_type = typename EvolvedVariableTag::type;
+  using error_variable_type =
+      typename db::add_tag_prefix<::Tags::StepperError,
+                                  EvolvedVariableTag>::type;
+
+  /// \cond
+  ErrorControl() = default;
+  explicit ErrorControl(CkMigrateMessage* /*unused*/) noexcept {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(ErrorControl);  // NOLINT
+  /// \endcond
+
+  static std::string name() noexcept {
+    if constexpr (std::is_same_v<ErrorControlSelector, NoSuchType>) {
+      return "ErrorControl";
+    } else {
+      return "ErrorControl(" + ErrorControlSelector::name() + ")";
+    }
+  }
+
+  struct AbsoluteTolerance {
+    using type = double;
+    static constexpr Options::String help{"Target absolute tolerance"};
+    static type lower_bound() noexcept { return 0.0; }
+  };
+
+  struct RelativeTolerance {
+    using type = double;
+    static constexpr Options::String help{"Target relative tolerance"};
+    static type lower_bound() noexcept { return 0.0; }
+  };
+
+  struct MaxFactor {
+    using type = double;
+    static constexpr Options::String help{
+        "Maximum factor to increase the step by"};
```

Comment:
(This is redundant with the Increase step chooser, but it seems to be considered part of the specific algorithm here so you can decide whether to keep it.)

Reply:
I think it's probably best to keep it in line with the Hairer explanation of the algorithm for now, so I'll keep the `MaxFactor`